azurelinuxagent/common/protocol/extensions_goal_state_from_vm_settings.py

```diff
@@ -325,7 +335,52 @@ def _parse_extensions(self, vm_settings):
                         raise Exception('additionalLocations should be an array')
                     extension.manifest_uris.extend(additional_locations)
 
-                # TODO: Add settings
+                #
+                # Settings
+                #
+                settings_list = extension_gs.get('settings')
+                if settings_list is not None:
+                    if not isinstance(settings_list, list):
+                        raise Exception("'settings' should be an array (extension: {0})".format(extension.name))
```

Comment:
I was wondering if we should raise these errors differently (like `VMSettingConfigError` or something) to have a logical separation of errors due to actual issues in the contract vs other unhandled exceptions. We don't need it now, but maybe later on we this separation would be helpful in some cases (like we have currently for `ExtensionConfigErrors` for detecting if last GS was unsupported or not

Reply:
All those exceptions are converted to a VmSettingsError by ExtensionsGoalStateFromVmSettings.__init__. We use these classes for bad goal states:
```

class ExtensionsGoalStateError(ExtensionError):
    """
    Error raised when the ExtensionsGoalState is malformed
    """


class ExtensionsConfigError(ExtensionsGoalStateError):
    """
    Error raised when the ExtensionsConfig is malformed
    """


class VmSettingsError(ExtensionsGoalStateError):
    """
    Error raised when the VmSettings are malformed
    """

```

Reply:
I saw that, let me rephrase, I meant separating out the errors where we were expecting something but it turned out to be something different vs something else that went wrong. Though I don't think there would be a big bucket of the latter but still I was wondering if having something like that would be useful.

Reply:
Not sure if that would be useful, I guess we could do it. __init__ also outputs the call stack so that would help debug issues other than parsing an incorrect goal state. Do you feel we need more than that?

Reply:
I think its fine for now, just wanted to brainstorm I guess a thought that came to mind